include/rcutils/error_handling.h

```diff
@@ -22,58 +22,119 @@ extern "C"
 {
 #endif
 
+#define __STDC_WANT_LIB_EXT1__ 1  // indicate we would like strnlen_s if available
+#include <assert.h>
 #include <stdbool.h>
 #include <stddef.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 
 #include "rcutils/allocator.h"
-#include "rcutils/format_string.h"
 #include "rcutils/macros.h"
+#include "rcutils/snprintf.h"
 #include "rcutils/types/rcutils_ret.h"
 #include "rcutils/visibility_control.h"
 
-/// Struct which encapsulates the error state set by RCUTILS_SET_ERROR_MSG().
-typedef struct rcutils_error_state_t
-{
-  const char * message;
-  const char * file;
-  size_t line_number;
-  rcutils_allocator_t allocator;
-} rcutils_error_state_t;
-
-// TODO(dhood): use __STDC_LIB_EXT1__ if/when supported in other implementations.
-#if defined(_WIN32)
+#ifdef __STDC_LIB_EXT1__
 // Limit the buffer size in the `fwrite` call to give an upper bound to buffer overrun in the case
 // of non-null terminated `msg`.
 #define RCUTILS_SAFE_FWRITE_TO_STDERR(msg) fwrite(msg, sizeof(char), strnlen_s(msg, 4096), stderr)
 #else
 #define RCUTILS_SAFE_FWRITE_TO_STDERR(msg) fwrite(msg, sizeof(char), strlen(msg), stderr)
 #endif
 
-/// Copy an error state into a destination error state.
+// fixed constraints
+#define RCUTILS_ERROR_STATE_LINE_NUMBER_STR_MAX_LENGTH 20  // "18446744073709551615"
+#define RCUTILS_ERROR_FORMATTING_CHARACTERS 6  // ', at ' + ':'
+
+// max formatted string length
+#define RCUTILS_ERROR_MESSAGE_MAX_LENGTH 1024
+
+// adjustable max length for user defined error message
+// remember "chained" errors will include previously specified file paths
+// e.g. "some error, at /path/to/a.c:42, at /path/to/b.c:42"
+#define RCUTILS_ERROR_STATE_MESSAGE_MAX_LENGTH 768
+// with RCUTILS_ERROR_STATE_MESSAGE_MAX_LENGTH = 768, RCUTILS_ERROR_STATE_FILE_MAX_LENGTH == 228
+#define RCUTILS_ERROR_STATE_FILE_MAX_LENGTH \
+  RCUTILS_ERROR_MESSAGE_MAX_LENGTH - \
+  RCUTILS_ERROR_STATE_MESSAGE_MAX_LENGTH - \
+  RCUTILS_ERROR_STATE_LINE_NUMBER_STR_MAX_LENGTH - \
+  RCUTILS_ERROR_FORMATTING_CHARACTERS - \
+  1
+
+/// Struct wrapping a fixed-size c string used for returning the formatted error string.
+typedef struct rcutils_error_string_t
+{
+  const char str[RCUTILS_ERROR_MESSAGE_MAX_LENGTH];
```

Comment:
Why should the member be `const`? Not making it const would:
* avoid the need to cast it away in functions like `__rcutils_format_error_string` and
* also match user expectations better imo.
If the variable of type `rcutils_error_string_t` itself is `const` then the member is const implicitly.

Same for the members of `rcutils_error_state_t` below.

Reply:
I can change them.